File: crates/axe-core/src/doctor.rs
//! `axe doctor`: a list of small read-only checks against a [`Layout`].
//!
//! Findings are intentionally tiny: a level, a topic, a one-line message.
//! Anything richer can wait for the v0.2 agent contract.

use std::fmt;

use serde::{Deserialize, Serialize};

use crate::Config;
use crate::layout::{Layout, Platform};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Level {
    Ok,
    Warn,
    Error,
}

impl fmt::Display for Level {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(match self {
            Self::Ok => "ok",
            Self::Warn => "warn",
            Self::Error => "error",
        })
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Finding {
    pub level: Level,
    pub topic: String,
    pub message: String,
}

impl Finding {
    fn ok(topic: impl Into<String>, message: impl Into<String>) -> Self {
        Self {
            level: Level::Ok,
            topic: topic.into(),
            message: message.into(),
        }
    }
    fn warn(topic: impl Into<String>, message: impl Into<String>) -> Self {
        Self {
            level: Level::Warn,
            topic: topic.into(),
            message: message.into(),
        }
    }
    fn error(topic: impl Into<String>, message: impl Into<String>) -> Self {
        Self {
            level: Level::Error,
            topic: topic.into(),
            message: message.into(),
        }
    }
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct Report {
    pub findings: Vec<Finding>,
}

impl Report {
    #[must_use]
    pub fn worst(&self) -> Level {
        self.findings
            .iter()
            .map(|f| f.level)
            .max_by_key(|l| match l {
                Level::Ok => 0,
                Level::Warn => 1,
                Level::Error => 2,
            })
            .unwrap_or(Level::Ok)
    }
}

/// Run all v0.1 checks against `layout` (and the config it was built from).
///
/// The checks are deliberately read-only: existence, executable bit, and
/// case-sensitivity sanity (Linux's lowercase `game.db` rule).
#[must_use]
pub fn run(config: &Config, layout: &Layout) -> Report {
    let mut findings = Vec::new();

    findings.push(check_root(layout));
    findings.push(check_binary(layout));
    findings.push(check_config_dir(layout));
    findings.extend(check_inis(layout));
    findings.push(check_modlist(layout));
    findings.push(check_workshop(layout));
    findings.push(check_game_db(layout));
    findings.push(check_logs_dir(layout));
    findings.push(check_server_id(config));

    Report { findings }
}

fn check_root(layout: &Layout) -> Finding {
    if layout.root.is_dir() {
        Finding::ok("root", format!("{}", layout.root))
    } else {
        Finding::error("root", format!("{} is not a directory", layout.root))
    }
}

fn check_binary(layout: &Layout) -> Finding {
    let bin = &layout.binary;
    if !bin.is_file() {
        return Finding::error("binary", format!("missing {bin}"));
    }
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        match std::fs::metadata(bin.as_std_path()) {
            Ok(m) if m.permissions().mode() & 0o111 != 0 => Finding::ok(
                "binary",
                format!("{} (executable)", bin.file_name().unwrap_or("?")),
            ),
            Ok(_) => Finding::warn("binary", format!("{bin} is not executable")),
            Err(e) => Finding::warn("binary", format!("stat {bin}: {e}")),
        }
    }
    #[cfg(not(unix))]
    {
        Finding::ok("binary", format!("{}", bin.file_name().unwrap_or("?")))
    }
}

fn check_config_dir(layout: &Layout) -> Finding {
    if layout.config_dir.is_dir() {
        Finding::ok("config", format!("{}", layout.config_dir))
    } else {
        Finding::warn(
            "config",
            format!(
                "{} not present yet (server has not generated configs)",
                layout.config_dir
            ),
        )
    }
}

fn check_inis(layout: &Layout) -> Vec<Finding> {
    [
        ("ServerSettings.ini", &layout.server_settings_ini),
        ("Engine.ini", &layout.engine_ini),
        ("Game.ini", &layout.game_ini),
    ]
    .into_iter()
    .map(|(label, path)| {
        if path.is_file() {
            Finding::ok("ini", format!("{label} present"))
        } else {
            Finding::warn("ini", format!("{label} not present at {path}"))
        }
    })
    .collect()
}

fn check_modlist(layout: &Layout) -> Finding {
    let path = &layout.modlist_txt;
    match std::fs::read_to_string(path.as_std_path()) {
        Ok(s) => {
            let n = s.lines().filter(|l| !l.trim().is_empty()).count();
            Finding::ok(
                "modlist",
                format!("{n} entr{} in {path}", if n == 1 { "y" } else { "ies" }),
            )
        }
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
            Finding::warn("modlist", format!("{path} not present (no mods enabled)"))
        }
        Err(e) => Finding::warn("modlist", format!("read {path}: {e}")),
    }
}

fn check_workshop(layout: &Layout) -> Finding {
    if layout.workshop_acf.is_file() {
        Finding::ok("workshop", format!("{} present", layout.workshop_acf))
    } else {
        Finding::warn(
            "workshop",
            format!(
                "{} not present (mods will populate it on first download)",
                layout.workshop_acf
            ),
        )
    }
}

fn check_game_db(layout: &Layout) -> Finding {
    let path = &layout.game_db;
    if path.is_file() {
        return Finding::ok("game.db", format!("{path}"));
    }
    if layout.platform != Platform::Linux {
        return Finding::warn("game.db", format!("{path} not present yet"));
    }
    // Linux is case-sensitive; warn if a wrongly-cased copy is sitting nearby.
    let saved = layout.root.join("ConanSandbox/Saved");
    let Ok(entries) = std::fs::read_dir(saved.as_std_path()) else {
        return Finding::warn("game.db", format!("{path} not present yet"));
    };
    for entry in entries.flatten() {
        let name = entry.file_name();
        let name = name.to_string_lossy();
        if name.eq_ignore_ascii_case("game.db") && name != "game.db" {
            return Finding::error(
                "game.db",
                format!("found {name} but Linux requires lowercase 'game.db'"),
            );
        }
    }
    Finding::warn("game.db", format!("{path} not present yet"))
}

fn check_logs_dir(layout: &Layout) -> Finding {
    if layout.logs_dir.is_dir() {
        Finding::ok("logs", format!("{}", layout.logs_dir))
    } else {
        Finding::warn("logs", format!("{} not present yet", layout.logs_dir))
    }
}

fn check_server_id(config: &Config) -> Finding {
    if config.server.id.trim().is_empty() {
        Finding::error("server.id", "axe.toml [server] id is empty")
    } else {
        Finding::ok("server.id", config.server.id.clone())
    }
}

#[cfg(test)]
mod tests {
    use std::fs;

    use camino::Utf8PathBuf;

    use super::*;
    use crate::layout::Layout;

    fn empty_install(platform: Platform) -> (tempfile::TempDir, Utf8PathBuf) {
        let dir = tempfile::tempdir().unwrap();
        let root: Utf8PathBuf = dir.path().to_path_buf().try_into().unwrap();
        let bin_dir = root.join(platform.binary_subdir());
        fs::create_dir_all(&bin_dir).unwrap();
        let bin = bin_dir.join(platform.binary_name());
        fs::write(&bin, b"").unwrap();
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let mut p = fs::metadata(&bin).unwrap().permissions();
            p.set_mode(0o755);
            fs::set_permissions(&bin, p).unwrap();
        }
        (dir, root)
    }

    #[test]
    fn fresh_linux_install_yields_warnings_only() {
        let (_g, root) = empty_install(Platform::Linux);
        let cfg = Config::from_root("test", root.clone());
        let layout = Layout::at(&root, Platform::Linux);
        let report = run(&cfg, &layout);
        assert_ne!(report.worst(), Level::Error);
    }

    #[test]
    fn detects_uppercase_game_db_on_linux() {
        let (_g, root) = empty_install(Platform::Linux);
        fs::create_dir_all(root.join("ConanSandbox/Saved").as_std_path()).unwrap();
        fs::write(root.join("ConanSandbox/Saved/Game.db").as_std_path(), b"").unwrap();
        let cfg = Config::from_root("test", root.clone());
        let layout = Layout::at(&root, Platform::Linux);
        let report = run(&cfg, &layout);
        assert!(
            report
                .findings
                .iter()
                .any(|f| f.topic == "game.db" && f.level == Level::Error)
        );
    }
}

File: src/core/doctor.ts
import { readdir, readFile, stat } from 'node:fs/promises';
import * as posix from 'node:path/posix';
import type { Config } from './config.ts';
import type { Layout } from './layout.ts';

export type Level = 'ok' | 'warn' | 'error';
export type Finding = { level: Level; topic: string; message: string };
export type Report = { findings: Finding[] };

/** Pure helper: returns the worst level present in the report. */
export function worstFinding(report: Report): Level {
  let worst: Level = 'ok';
  for (const f of report.findings) {
    if (f.level === 'error') return 'error';
    if (f.level === 'warn' && worst === 'ok') worst = 'warn';
  }
  return worst;
}

/** Run all v0.1 read-only checks against `layout` and the config it came from. */
export async function runDoctor(config: Config, layout: Layout): Promise<Report> {
  const findings: Finding[] = [];
  findings.push(await checkRoot(layout));
  findings.push(await checkBinary(layout));
  findings.push(await checkConfigDir(layout));
  findings.push(...(await checkInis(layout)));
  findings.push(await checkModlist(layout));
  findings.push(await checkWorkshop(layout));
  findings.push(await checkGameDb(layout));
  findings.push(await checkLogsDir(layout));
  findings.push(checkServerId(config));
  return { findings };
}

async function checkRoot(layout: Layout): Promise<Finding> {
  if (await isDirectory(layout.root)) {
    return { level: 'ok', topic: 'root', message: layout.root };
  }
  return { level: 'error', topic: 'root', message: `${layout.root} is not a directory` };
}

async function checkBinary(layout: Layout): Promise<Finding> {
  if (!(await isFile(layout.binary))) {
    return { level: 'error', topic: 'binary', message: `missing ${layout.binary}` };
  }
  if (process.platform !== 'win32') {
    try {
      const st = await stat(layout.binary);
      const executable = (st.mode & 0o111) !== 0;
      const name = posix.basename(layout.binary);
      if (executable) {
        return { level: 'ok', topic: 'binary', message: `${name} (executable)` };
      }
      return { level: 'warn', topic: 'binary', message: `${layout.binary} is not executable` };
    } catch (e) {
      return {
        level: 'warn',
        topic: 'binary',
        message: `stat ${layout.binary}: ${(e as Error).message}`,
      };
    }
  }
  return { level: 'ok', topic: 'binary', message: posix.basename(layout.binary) };
}

async function checkConfigDir(layout: Layout): Promise<Finding> {
  if (await isDirectory(layout.config_dir)) {
    return { level: 'ok', topic: 'config', message: layout.config_dir };
  }
  return {
    level: 'warn',
    topic: 'config',
    message: `${layout.config_dir} not present yet (server has not generated configs)`,
  };
}

async function checkInis(layout: Layout): Promise<Finding[]> {
  const checks: Array<[string, string]> = [
    ['ServerSettings.ini', layout.server_settings_ini],
    ['Engine.ini', layout.engine_ini],
    ['Game.ini', layout.game_ini],
  ];
  const out: Finding[] = [];
  for (const [label, path] of checks) {
    if (await isFile(path)) {
      out.push({ level: 'ok', topic: 'ini', message: `${label} present` });
    } else {
      out.push({ level: 'warn', topic: 'ini', message: `${label} not present at ${path}` });
    }
  }
  return out;
}

async function checkModlist(layout: Layout): Promise<Finding> {
  try {
    const text = await readFile(layout.modlist_txt, 'utf8');
    const n = text.split('\n').filter((l) => l.trim().length > 0).length;
    return {
      level: 'ok',
      topic: 'modlist',
      message: `${n} entr${n === 1 ? 'y' : 'ies'} in ${layout.modlist_txt}`,
    };
  } catch (e) {
    const code = (e as NodeJS.ErrnoException).code;
    if (code === 'ENOENT') {
      return {
        level: 'warn',
        topic: 'modlist',
        message: `${layout.modlist_txt} not present (no mods enabled)`,
      };
    }
    return {
      level: 'warn',
      topic: 'modlist',
      message: `read ${layout.modlist_txt}: ${(e as Error).message}`,
    };
  }
}

async function checkWorkshop(layout: Layout): Promise<Finding> {
  if (await isFile(layout.workshop_acf)) {
    return { level: 'ok', topic: 'workshop', message: `${layout.workshop_acf} present` };
  }
  return {
    level: 'warn',
    topic: 'workshop',
    message: `${layout.workshop_acf} not present (mods will populate it on first download)`,
  };
}

async function checkGameDb(layout: Layout): Promise<Finding> {
  if (await isFile(layout.game_db)) {
    return { level: 'ok', topic: 'game.db', message: layout.game_db };
  }
  if (layout.platform !== 'linux') {
    return { level: 'warn', topic: 'game.db', message: `${layout.game_db} not present yet` };
  }
  // Linux is case-sensitive; warn if a mis-cased file exists.
  const saved = posix.join(layout.root, 'ConanSandbox/Saved');
  try {
    const entries = await readdir(saved);
    for (const name of entries) {
      if (name.toLowerCase() === 'game.db' && name !== 'game.db') {
        return {
          level: 'error',
          topic: 'game.db',
          message: `found ${name} but Linux requires lowercase 'game.db'`,
        };
      }
    }
  } catch {
    /* saved/ doesn't exist yet — fall through */
  }
  return { level: 'warn', topic: 'game.db', message: `${layout.game_db} not present yet` };
}

async function checkLogsDir(layout: Layout): Promise<Finding> {
  if (await isDirectory(layout.logs_dir)) {
    return { level: 'ok', topic: 'logs', message: layout.logs_dir };
  }
  return { level: 'warn', topic: 'logs', message: `${layout.logs_dir} not present yet` };
}

function checkServerId(config: Config): Finding {
  if (config.server.id.trim().length === 0) {
    return { level: 'error', topic: 'server.id', message: 'axe.toml [server] id is empty' };
  }
  return { level: 'ok', topic: 'server.id', message: config.server.id };
}

async function isDirectory(p: string): Promise<boolean> {
  try {
    return (await stat(p)).isDirectory();
  } catch {
    return false;
  }
}

async function isFile(p: string): Promise<boolean> {
  try {
    return (await stat(p)).isFile();
  } catch {
    return false;
  }
}
